// qaseio/dist/src/model/filters1.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * Qase.io API
 * Qase API Specification.
 *
 * The version of the OpenAPI document: 1.0.0
 * Contact: support@qase.io
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.Filters1TypeEnum = exports.Filters1EntityEnum = void 0;
/**
    * @export
    * @enum {string}
    */
var Filters1EntityEnum;
(function (Filters1EntityEnum) {
    Filters1EntityEnum["CASE"] = "case";
    Filters1EntityEnum["RUN"] = "run";
    Filters1EntityEnum["DEFECT"] = "defect";
})(Filters1EntityEnum = exports.Filters1EntityEnum || (exports.Filters1EntityEnum = {}));
/**
    * @export
    * @enum {string}
    */
var Filters1TypeEnum;
(function (Filters1TypeEnum) {
    Filters1TypeEnum["STRING"] = "string";
    Filters1TypeEnum["TEXT"] = "text";
    Filters1TypeEnum["NUMBER"] = "number";
    Filters1TypeEnum["CHECKBOX"] = "checkbox";
    Filters1TypeEnum["SELECTBOX"] = "selectbox";
    Filters1TypeEnum["RADIO"] = "radio";
    Filters1TypeEnum["MULTISELECT"] = "multiselect";
    Filters1TypeEnum["URL"] = "url";
    Filters1TypeEnum["USER"] = "user";
    Filters1TypeEnum["DATETIME"] = "datetime";
})(Filters1TypeEnum = exports.Filters1TypeEnum || (exports.Filters1TypeEnum = {}));
