// playwright-qase-reporter/dist/index.js
"use strict";
/* eslint-disable max-len */
/* eslint-disable camelcase */
/* eslint-disable @typescript-eslint/no-unsafe-member-access */
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __makeTemplateObject = (this && this.__makeTemplateObject) || function (cooked, raw) {
    if (Object.defineProperty) { Object.defineProperty(cooked, "raw", { value: raw }); } else { cooked.raw = raw; }
    return cooked;
};
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __spreadArrays = (this && this.__spreadArrays) || function () {
    for (var s = 0, i = 0, il = arguments.length; i < il; i++) s += arguments[i].length;
    for (var r = Array(s), k = 0, i = 0; i < il; i++)
        for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
            r[k] = a[j];
    return r;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var src_1 = require("qaseio/dist/src");
var fs_1 = require("fs");
var form_data_1 = __importDefault(require("form-data"));
var qaseio_1 = require("qaseio");
var chalk_1 = __importDefault(require("chalk"));
var crypto_1 = __importDefault(require("crypto"));
var child_process_1 = require("child_process");
var Envs;
(function (Envs) {
    Envs["report"] = "QASE_REPORT";
    Envs["apiToken"] = "QASE_API_TOKEN";
    Envs["basePath"] = "QASE_API_BASE_URL";
    Envs["rootSuiteTitle"] = "QASE_ROOT_SUITE_TITLE";
    Envs["projectCode"] = "QASE_PROJECT_CODE";
    Envs["runId"] = "QASE_RUN_ID";
    Envs["runName"] = "QASE_RUN_NAME";
    Envs["runDescription"] = "QASE_RUN_DESCRIPTION";
    Envs["runComplete"] = "QASE_RUN_COMPLETE";
    Envs["environmentId"] = "QASE_ENVIRONMENT_ID";
    Envs["uploadAttachments"] = "QASE_UPLOAD_ATTACHMENTS";
})(Envs || (Envs = {}));
var Statuses = {
    passed: src_1.ResultCreateStatusEnum.PASSED,
    failed: src_1.ResultCreateStatusEnum.FAILED,
    skipped: src_1.ResultCreateStatusEnum.SKIPPED,
    pending: src_1.ResultCreateStatusEnum.SKIPPED,
    disabled: src_1.ResultCreateStatusEnum.BLOCKED,
};
var customBoundary = '----------------------------';
crypto_1.default.randomBytes(24).forEach(function (value) {
    customBoundary += Math.floor(value * 10).toString(16);
});
var CustomBoundaryFormData = /** @class */ (function (_super) {
    __extends(CustomBoundaryFormData, _super);
    function CustomBoundaryFormData() {
        return _super.call(this) || this;
    }
    CustomBoundaryFormData.prototype.getBoundary = function () {
        return customBoundary;
    };
    return CustomBoundaryFormData;
}(form_data_1.default));
var PlaywrightReporter = /** @class */ (function () {
    function PlaywrightReporter(_options) {
        this.isDisabled = false;
        this.queued = 0;
        this.resultsToBePublished = [];
        this.options = _options;
        this.options.runComplete = !!PlaywrightReporter.getEnv(Envs.runComplete) || this.options.runComplete;
        this.options.uploadAttachments = !!PlaywrightReporter.getEnv(Envs.uploadAttachments) || this.options.uploadAttachments;
        this.options.rootSuiteTitle = _options.rootSuiteTitle || PlaywrightReporter.getEnv(Envs.rootSuiteTitle) || '';
        this.options.projectCode = _options.projectCode || PlaywrightReporter.getEnv(Envs.projectCode) || '';
        this.api = new qaseio_1.QaseApi(PlaywrightReporter.getEnv(Envs.apiToken) || this.options.apiToken || '', PlaywrightReporter.getEnv(Envs.basePath) || this.options.basePath, PlaywrightReporter.createHeaders(), CustomBoundaryFormData);
        this.log(chalk_1.default(templateObject_1 || (templateObject_1 = __makeTemplateObject(["{yellow Current PID: ", "}"], ["{yellow Current PID: ", "}"])), process.pid));
        if (!PlaywrightReporter.getEnv(Envs.report)) {
            this.log(chalk_1.default(templateObject_2 || (templateObject_2 = __makeTemplateObject(["{yellow QASE_REPORT env variable is not set. Reporting to qase.io is disabled.}"], ["{yellow QASE_REPORT env variable is not set. Reporting to qase.io is disabled.}"]))));
            this.isDisabled = true;
            return;
        }
    }
    PlaywrightReporter.createRunObject = function (name, cases, args) {
        return __assign({ title: name, cases: cases }, args);
    };
    PlaywrightReporter.getEnv = function (name) {
        return process.env[name];
    };
    PlaywrightReporter.getPackageVersion = function (name) {
        var UNDEFINED = 'undefined';
        try {
            var pathToPackageJson = require.resolve(name + "/package.json", { paths: [process.cwd()] });
            if (pathToPackageJson) {
                try {
                    var packageString = fs_1.readFileSync(pathToPackageJson, { encoding: 'utf8' });
                    if (packageString) {
                        var packageObject = JSON.parse(packageString);
                        return packageObject.version;
                    }
                    return UNDEFINED;
                }
                catch (error) {
                    return UNDEFINED;
                }
            }
        }
        catch (error) {
            return UNDEFINED;
        }
    };
    PlaywrightReporter.createHeaders = function () {
        var nodeVersion = process.version, os = process.platform, arch = process.arch;
        var npmVersion = child_process_1.execSync('npm -v', { encoding: 'utf8' }).replace(/['"\n]+/g, '');
        var qaseapiVersion = PlaywrightReporter.getPackageVersion('qaseio');
        var playwrightVersion = PlaywrightReporter.getPackageVersion('playwright-core');
        var playwrightCaseReporterVersion = PlaywrightReporter.getPackageVersion('playwright-qase-reporter');
        var xPlatformHeader = "node=" + nodeVersion + "; npm=" + npmVersion + "; os=" + os + "; arch=" + arch;
        var xClientHeader = "playwright=" + playwrightVersion + "; qase-playwright=" + playwrightCaseReporterVersion + "; qaseapi=" + qaseapiVersion;
        return {
            'X-Client': xClientHeader,
            'X-Platform': xPlatformHeader,
        };
    };
    PlaywrightReporter.getSuitePath = function (suite) {
        if (suite.parent) {
            var parentSuite = String(PlaywrightReporter.getSuitePath(suite.parent));
            if (parentSuite) {
                return parentSuite + '\t' + String(suite === null || suite === void 0 ? void 0 : suite.title);
            }
            else {
                return String(suite === null || suite === void 0 ? void 0 : suite.title);
            }
        }
        // eslint-disable-next-line @typescript-eslint/no-unsafe-return
        return suite.title;
    };
    PlaywrightReporter.prototype.onBegin = function () {
        return __awaiter(this, void 0, void 0, function () {
            var _this = this;
            return __generator(this, function (_a) {
                if (this.isDisabled) {
                    return [2 /*return*/];
                }
                return [2 /*return*/, this.checkProject(this.options.projectCode, function (prjExists) { return __awaiter(_this, void 0, void 0, function () {
                        var userDefinedRunId;
                        var _this = this;
                        return __generator(this, function (_a) {
                            if (!prjExists) {
                                this.log(chalk_1.default(templateObject_3 || (templateObject_3 = __makeTemplateObject(["{red Project ", " does not exist}"], ["{red Project ", " does not exist}"])), this.options.projectCode));
                                this.isDisabled = true;
                                return [2 /*return*/];
                            }
                            this.log(chalk_1.default(templateObject_4 || (templateObject_4 = __makeTemplateObject(["{green Project ", " exists}"], ["{green Project ", " exists}"])), this.options.projectCode));
                            userDefinedRunId = PlaywrightReporter.getEnv(Envs.runId) || this.options.runId;
                            if (userDefinedRunId) {
                                this.runId = userDefinedRunId;
                                return [2 /*return*/, this.checkRun(this.runId, function (runExists) {
                                        if (runExists) {
                                            _this.log(chalk_1.default(templateObject_5 || (templateObject_5 = __makeTemplateObject(["{green Using run ", " to publish test results}"], ["{green Using run ", " to publish test results}"])), _this.runId));
                                        }
                                        else {
                                            _this.log(chalk_1.default(templateObject_6 || (templateObject_6 = __makeTemplateObject(["{red Run ", " does not exist}"], ["{red Run ", " does not exist}"])), _this.runId));
                                            _this.isDisabled = true;
                                        }
                                    })];
                            }
                            else {
                                return [2 /*return*/, this.createRun(PlaywrightReporter.getEnv(Envs.runName), PlaywrightReporter.getEnv(Envs.runDescription), function (created) {
                                        var _a;
                                        if (created) {
                                            _this.runId = (_a = created.result) === null || _a === void 0 ? void 0 : _a.id;
                                            process.env.QASE_RUN_ID = String(_this === null || _this === void 0 ? void 0 : _this.runId);
                                            _this.log(chalk_1.default(templateObject_7 || (templateObject_7 = __makeTemplateObject(["{green Using run ", " to publish test results}"], ["{green Using run ", " to publish test results}"])), _this.runId));
                                        }
                                        else {
                                            _this.log(chalk_1.default(templateObject_8 || (templateObject_8 = __makeTemplateObject(["{red Could not create run in project ", "}"], ["{red Could not create run in project ", "}"])), _this.options.projectCode));
                                            _this.isDisabled = true;
                                        }
                                    })];
                            }
                            return [2 /*return*/];
                        });
                    }); })];
            });
        });
    };
    PlaywrightReporter.prototype.onTestEnd = function (test, testResult) {
        return __awaiter(this, void 0, void 0, function () {
            var attachmentsArray;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        if (this.isDisabled) {
                            return [2 /*return*/];
                        }
                        this.queued++;
                        attachmentsArray = [];
                        if (!(this.options.uploadAttachments && testResult.attachments.length > 0)) return [3 /*break*/, 2];
                        return [4 /*yield*/, this.uploadAttachments(testResult)];
                    case 1:
                        attachmentsArray = _a.sent();
                        _a.label = 2;
                    case 2: return [2 /*return*/, this.prepareCaseResult(test, testResult, attachmentsArray)];
                }
            });
        });
    };
    PlaywrightReporter.prototype.onEnd = function () {
        var _a, _b, _c, _d, _e, _f;
        return __awaiter(this, void 0, void 0, function () {
            var body, err_1, error_1, err;
            var _this = this;
            return __generator(this, function (_g) {
                switch (_g.label) {
                    case 0:
                        if (this.isDisabled) {
                            return [2 /*return*/];
                        }
                        return [4 /*yield*/, new Promise(function (resolve, reject) {
                                var timer = 0;
                                var interval = setInterval(function () {
                                    timer++;
                                    if (_this.runId && _this.queued === 0) {
                                        clearInterval(interval);
                                        resolve();
                                    }
                                    if (timer > 30) {
                                        clearInterval(interval);
                                        reject();
                                    }
                                }, 1000);
                            })];
                    case 1:
                        _g.sent();
                        if (this.resultsToBePublished.length === 0) {
                            this.log('No testcases were matched. Ensure that your tests are declared correctly.');
                            return [2 /*return*/];
                        }
                        body = {
                            results: this.resultsToBePublished,
                        };
                        _g.label = 2;
                    case 2:
                        _g.trys.push([2, 8, , 9]);
                        return [4 /*yield*/, this.api.results.createResultBulk(this.options.projectCode, Number(this.runId), body)];
                    case 3:
                        _g.sent();
                        this.log(chalk_1.default(templateObject_9 || (templateObject_9 = __makeTemplateObject(["{green ", " result(s) sent to Qase}"], ["{green ", " result(s) sent to Qase}"])), this.resultsToBePublished.length));
                        if (!this.options.runComplete) {
                            return [2 /*return*/];
                        }
                        _g.label = 4;
                    case 4:
                        _g.trys.push([4, 6, , 7]);
                        return [4 /*yield*/, this.api.runs.completeRun(this.options.projectCode, Number(this.runId))];
                    case 5:
                        _g.sent();
                        this.log(chalk_1.default(templateObject_10 || (templateObject_10 = __makeTemplateObject(["{green Run ", " completed}"], ["{green Run ", " completed}"])), this.runId));
                        return [3 /*break*/, 7];
                    case 6:
                        err_1 = _g.sent();
                        this.log("Error on completing run " + err_1);
                        return [3 /*break*/, 7];
                    case 7:
                        this.log(chalk_1.default(templateObject_11 || (templateObject_11 = __makeTemplateObject(["{blue https://app.qase.io/run/", "/dashboard/", "}"], ["{blue https://app.qase.io/run/", "/dashboard/", "}"])), this.options.projectCode, this.runId));
                        return [3 /*break*/, 9];
                    case 8:
                        error_1 = _g.sent();
                        err = error_1;
                        this.log(chalk_1.default(templateObject_12 || (templateObject_12 = __makeTemplateObject(["{red Unable to send results into Qase. ", "}"], ["{red Unable to send results into Qase. ", "}"])), err));
                        if (((_a = err === null || err === void 0 ? void 0 : err.response) === null || _a === void 0 ? void 0 : _a.status) >= 400 && ((_b = err === null || err === void 0 ? void 0 : err.response) === null || _b === void 0 ? void 0 : _b.status) < 500) {
                            this.log(chalk_1.default(templateObject_13 || (templateObject_13 = __makeTemplateObject(["{red Erorr message: ", "}"], ["{red Erorr message: ", "}"])), (_d = (_c = err === null || err === void 0 ? void 0 : err.response) === null || _c === void 0 ? void 0 : _c.data) === null || _d === void 0 ? void 0 : _d.errorMessage));
                            this.log(chalk_1.default(templateObject_14 || (templateObject_14 = __makeTemplateObject(["{red Erorrs: ", "}"], ["{red Erorrs: ", "}"])), JSON.stringify((_f = (_e = err === null || err === void 0 ? void 0 : err.response) === null || _e === void 0 ? void 0 : _e.data) === null || _f === void 0 ? void 0 : _f.errorFields)));
                        }
                        return [3 /*break*/, 9];
                    case 9: return [2 /*return*/];
                }
            });
        });
    };
    PlaywrightReporter.prototype.log = function (message) {
        var optionalParams = [];
        for (var _i = 1; _i < arguments.length; _i++) {
            optionalParams[_i - 1] = arguments[_i];
        }
        if (this.options.logging) {
            // eslint-disable-next-line no-console
            console.log.apply(console, __spreadArrays([chalk_1.default(templateObject_15 || (templateObject_15 = __makeTemplateObject(["{bold {blue qase:}} ", ""], ["{bold {blue qase:}} ", ""])), message)], optionalParams));
        }
    };
    PlaywrightReporter.prototype.getCaseIds = function (test) {
        var regexp = /(\(Qase ID: ([\d,]+)\))/;
        var results = regexp.exec(test.title);
        if (results && results.length === 3) {
            return results[2].split(',').map(function (value) { return Number.parseInt(value, 10); });
        }
        return [];
    };
    PlaywrightReporter.prototype.logTestItem = function (test, testResult) {
        var map = {
            failed: chalk_1.default(templateObject_16 || (templateObject_16 = __makeTemplateObject(["{red Test ", " ", "}"], ["{red Test ", " ", "}"])), test.title, testResult.status),
            passed: chalk_1.default(templateObject_17 || (templateObject_17 = __makeTemplateObject(["{green Test ", " ", "}"], ["{green Test ", " ", "}"])), test.title, testResult.status),
            skipped: chalk_1.default(templateObject_18 || (templateObject_18 = __makeTemplateObject(["{blueBright Test ", " ", "}"], ["{blueBright Test ", " ", "}"])), test.title, testResult.status),
            pending: chalk_1.default(templateObject_19 || (templateObject_19 = __makeTemplateObject(["{blueBright Test ", " ", "}"], ["{blueBright Test ", " ", "}"])), test.title, testResult.status),
            disabled: chalk_1.default(templateObject_20 || (templateObject_20 = __makeTemplateObject(["{gray Test ", " ", "}"], ["{gray Test ", " ", "}"])), test.title, testResult.status),
        };
        if (testResult.status) {
            this.log(map[testResult.status]);
        }
    };
    PlaywrightReporter.prototype.checkProject = function (projectCode, cb) {
        var _a;
        return __awaiter(this, void 0, void 0, function () {
            var response, err_2;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        _b.trys.push([0, 3, , 4]);
                        return [4 /*yield*/, this.api.projects.getProject(projectCode)];
                    case 1:
                        response = _b.sent();
                        return [4 /*yield*/, cb(Boolean((_a = response.data.result) === null || _a === void 0 ? void 0 : _a.code))];
                    case 2:
                        _b.sent();
                        return [3 /*break*/, 4];
                    case 3:
                        err_2 = _b.sent();
                        this.log(err_2);
                        this.isDisabled = true;
                        return [3 /*break*/, 4];
                    case 4: return [2 /*return*/];
                }
            });
        });
    };
    PlaywrightReporter.prototype.createRun = function (name, description, cb) {
        return __awaiter(this, void 0, void 0, function () {
            var environmentId, runObject, res, err_3;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        _a.trys.push([0, 2, , 3]);
                        environmentId = Number.parseInt(PlaywrightReporter.getEnv(Envs.environmentId), 10) || this.options.environmentId;
                        runObject = PlaywrightReporter.createRunObject(name || "Automated run " + new Date().toISOString(), [], {
                            description: description || 'Playwright automated run',
                            environment_id: environmentId,
                            is_autotest: true,
                        });
                        return [4 /*yield*/, this.api.runs.createRun(this.options.projectCode, runObject)];
                    case 1:
                        res = _a.sent();
                        cb(res.data);
                        return [3 /*break*/, 3];
                    case 2:
                        err_3 = _a.sent();
                        this.log("Error on creating run " + err_3);
                        this.isDisabled = true;
                        return [3 /*break*/, 3];
                    case 3: return [2 /*return*/];
                }
            });
        });
    };
    PlaywrightReporter.prototype.checkRun = function (runId, cb) {
        return __awaiter(this, void 0, void 0, function () {
            var _this = this;
            return __generator(this, function (_a) {
                if (runId === undefined) {
                    cb(false);
                    return [2 /*return*/];
                }
                return [2 /*return*/, this.api.runs.getRun(this.options.projectCode, Number(runId))
                        .then(function (resp) {
                        var _a, _b;
                        _this.log("Get run result on checking run " + ((_a = resp.data.result) === null || _a === void 0 ? void 0 : _a.id));
                        cb(Boolean((_b = resp.data.result) === null || _b === void 0 ? void 0 : _b.id));
                    })
                        .catch(function (err) {
                        _this.log("Error on checking run " + err);
                        _this.isDisabled = true;
                    })];
            });
        });
    };
    PlaywrightReporter.prototype.uploadAttachments = function (testResult) {
        return __awaiter(this, void 0, void 0, function () {
            var _this = this;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, Promise.all(testResult.attachments.map(function (attachment) { return __awaiter(_this, void 0, void 0, function () {
                            var data, options, response;
                            var _a;
                            return __generator(this, function (_b) {
                                switch (_b.label) {
                                    case 0:
                                        data = fs_1.createReadStream(attachment === null || attachment === void 0 ? void 0 : attachment.path);
                                        options = {
                                            headers: {
                                                'Content-Type': 'multipart/form-data; boundary=' + customBoundary,
                                            },
                                        };
                                        return [4 /*yield*/, this.api.attachments.uploadAttachment(this.options.projectCode, [data], options)];
                                    case 1:
                                        response = _b.sent();
                                        // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
                                        return [2 /*return*/, (_a = response.data.result) === null || _a === void 0 ? void 0 : _a[0].hash];
                                }
                            });
                        }); }))];
                    case 1: return [2 /*return*/, _a.sent()];
                }
            });
        });
    };
    PlaywrightReporter.prototype.removePublished = function (testAlias) {
        var resultIndex = this.resultsToBePublished.indexOf(testAlias);
        if (resultIndex !== -1) {
            this.resultsToBePublished.splice(resultIndex, 1);
        }
    };
    PlaywrightReporter.prototype.prepareCaseResult = function (test, testResult, attachments) {
        var _this = this;
        var _a, _b, _c, _d;
        this.queued--;
        this.logTestItem(test, testResult);
        var caseIds = this.getCaseIds(test);
        var caseObject = {
            // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
            status: Statuses[testResult.status] || Statuses.failed,
            time_ms: testResult.duration,
            stacktrace: (_b = (_a = testResult.error) === null || _a === void 0 ? void 0 : _a.stack) === null || _b === void 0 ? void 0 : _b.replace(/\u001b\[.*?m/g, ''),
            comment: testResult.error ? test.title + ": " + ((_d = (_c = testResult.error) === null || _c === void 0 ? void 0 : _c.message) === null || _d === void 0 ? void 0 : _d.replace(/\u001b\[.*?m/g, '')) : undefined,
            attachments: attachments.length > 0
                ? attachments
                : undefined,
            defect: Statuses[testResult.status] === Statuses.failed,
        };
        if (caseIds.length === 0) {
            caseObject.case = {
                title: test.title,
                suite_title: this.options.rootSuiteTitle
                    ? this.options.rootSuiteTitle + "\t" + PlaywrightReporter.getSuitePath(test.parent)
                    : PlaywrightReporter.getSuitePath(test.parent),
            };
            this.resultsToBePublished.push(caseObject);
            this.log(chalk_1.default(templateObject_21 || (templateObject_21 = __makeTemplateObject(["{gray Result prepared for publish: ", " }"], ["{gray Result prepared for publish: ", " }"])), test.title));
        }
        else {
            caseIds.forEach(function (caseId) {
                var add = caseIds.length > 1 ? chalk_1.default(templateObject_22 || (templateObject_22 = __makeTemplateObject([" {white For case ", "}"], [" {white For case ", "}"])), caseId) : '';
                _this.log(chalk_1.default(templateObject_23 || (templateObject_23 = __makeTemplateObject(["{gray Start publishing: ", "}", ""], ["{gray Start publishing: ", "}", ""])), test.title, add));
                var caseObjectWithId = __assign({ case_id: caseId }, caseObject);
                _this.resultsToBePublished.push(caseObjectWithId);
                _this.log(chalk_1.default(templateObject_24 || (templateObject_24 = __makeTemplateObject(["{gray Result prepared for publish: ", " }", ""], ["{gray Result prepared for publish: ", " }", ""])), test.title, add));
            });
        }
    };
    return PlaywrightReporter;
}());
exports.default = PlaywrightReporter;
var templateObject_1, templateObject_2, templateObject_3, templateObject_4, templateObject_5, templateObject_6, templateObject_7, templateObject_8, templateObject_9, templateObject_10, templateObject_11, templateObject_12, templateObject_13, templateObject_14, templateObject_15, templateObject_16, templateObject_17, templateObject_18, templateObject_19, templateObject_20, templateObject_21, templateObject_22, templateObject_23, templateObject_24;
