// contentful-management/dist/es-modules/entities/app-upload.js
import { freezeSys, toPlainObject } from 'contentful-sdk-core';
import copy from 'fast-copy';
import { wrapCollection } from '../common-utils';
import enhanceWithMethods from '../enhance-with-methods';
/**
 * @private
 */
function createAppUploadApi(makeRequest) {
  const getParams = data => ({
    organizationId: data.sys.organization.sys.id,
    appUploadId: data.sys.id
  });
  return {
    delete: function del() {
      const data = this.toPlainObject();
      return makeRequest({
        entityType: 'AppUpload',
        action: 'delete',
        params: getParams(data)
      });
    }
  };
}

/**
 * @private
 * @param makeRequest - function to make requests via an adapter
 * @param data - Raw App Upload data
 * @return Wrapped App Upload data
 */
export function wrapAppUpload(makeRequest, data) {
  const appUpload = toPlainObject(copy(data));
  const appUploadWithMethods = enhanceWithMethods(appUpload, createAppUploadApi(makeRequest));
  return freezeSys(appUploadWithMethods);
}

/**
 * @private
 * @param makeRequest - function to make requests via an adapter
 * @param data - Raw App Upload collection data
 * @return Wrapped App Upload collection data
 */
export const wrapAppUploadCollection = wrapCollection(wrapAppUpload);
