// contentful-management/dist/es-modules/entities/app-details.js
import copy from 'fast-copy';
import { toPlainObject } from 'contentful-sdk-core';
import enhanceWithMethods from '../enhance-with-methods';
export let IconType = /*#__PURE__*/function (IconType) {
  IconType["Base64"] = "base64";
  return IconType;
}({});
/**
 * @private
 * @param makeRequest - function to make requests via an adapter
 * @return Wrapped App Details data
 */
function createAppDetailsApi(makeRequest) {
  const getParams = data => ({
    organizationId: data.sys.organization.sys.id,
    appDefinitionId: data.sys.appDefinition.sys.id
  });
  return {
    delete: function del() {
      const self = this;
      return makeRequest({
        entityType: 'AppDetails',
        action: 'delete',
        params: getParams(self)
      });
    }
  };
}

/**
 * @private
 * @param http - HTTP client instance
 * @param data - Raw AppDetails data
 * @return Wrapped AppDetails data
 */
export function wrapAppDetails(makeRequest, data) {
  const appDetails = toPlainObject(copy(data));
  return enhanceWithMethods(appDetails, createAppDetailsApi(makeRequest));
}
