// contentful-management/dist/es-modules/entities/environment-template-installation.js
import { freezeSys, toPlainObject } from 'contentful-sdk-core';
import copy from 'fast-copy';
import { wrapCursorPaginatedCollection } from '../common-utils';
var EnvironmentTemplateInstallationStatuses = /*#__PURE__*/function (EnvironmentTemplateInstallationStatuses) {
  EnvironmentTemplateInstallationStatuses["created"] = "created";
  EnvironmentTemplateInstallationStatuses["inProgress"] = "inProgress";
  EnvironmentTemplateInstallationStatuses["failed"] = "failed";
  EnvironmentTemplateInstallationStatuses["succeeded"] = "succeeded";
  EnvironmentTemplateInstallationStatuses["disconnected"] = "disconnected";
  EnvironmentTemplateInstallationStatuses["inRetry"] = "inRetry";
  return EnvironmentTemplateInstallationStatuses;
}(EnvironmentTemplateInstallationStatuses || {});
export function wrapEnvironmentTemplateInstallation(makeRequest, data) {
  const environmentTemplate = toPlainObject(copy(data));
  return freezeSys(environmentTemplate);
}
export const wrapEnvironmentTemplateInstallationCollection = wrapCursorPaginatedCollection(wrapEnvironmentTemplateInstallation);
