// contentful-management/dist/es-modules/entities/organization-membership.js
import copy from 'fast-copy';
import { freezeSys, toPlainObject } from 'contentful-sdk-core';
import enhanceWithMethods from '../enhance-with-methods';
import { wrapCollection } from '../common-utils';
/**
 * @private
 */
function createOrganizationMembershipApi(makeRequest, organizationId) {
  const getParams = data => ({
    organizationMembershipId: data.sys.id,
    organizationId
  });
  return {
    update: function update() {
      const raw = this.toPlainObject();
      return makeRequest({
        entityType: 'OrganizationMembership',
        action: 'update',
        params: getParams(raw),
        payload: raw
      }).then(data => wrapOrganizationMembership(makeRequest, data, organizationId));
    },
    delete: function del() {
      const raw = this.toPlainObject();
      return makeRequest({
        entityType: 'OrganizationMembership',
        action: 'delete',
        params: getParams(raw)
      });
    }
  };
}

/**
 * @private
 * @param {function} makeRequest - function to make requests via an adapter
 * @param {Object} data - Raw organization membership data
 * @return {OrganizationMembership} Wrapped organization membership data
 */
export function wrapOrganizationMembership(makeRequest, data, organizationId) {
  const organizationMembership = toPlainObject(copy(data));
  const organizationMembershipWithMethods = enhanceWithMethods(organizationMembership, createOrganizationMembershipApi(makeRequest, organizationId));
  return freezeSys(organizationMembershipWithMethods);
}

/**
 * @private
 */
export const wrapOrganizationMembershipCollection = wrapCollection(wrapOrganizationMembership);
