// contentful-management/dist/es-modules/entities/role.js
import copy from 'fast-copy';
import { freezeSys, toPlainObject } from 'contentful-sdk-core';
import enhanceWithMethods from '../enhance-with-methods';
import { wrapCollection } from '../common-utils';
/**
 * @private
 */
function createRoleApi(makeRequest) {
  const getParams = data => ({
    spaceId: data.sys.space.sys.id,
    roleId: data.sys.id
  });
  return {
    update: function update() {
      const data = this.toPlainObject();
      return makeRequest({
        entityType: 'Role',
        action: 'update',
        params: getParams(data),
        payload: data
      }).then(data => wrapRole(makeRequest, data));
    },
    delete: function del() {
      const data = this.toPlainObject();
      return makeRequest({
        entityType: 'Role',
        action: 'delete',
        params: getParams(data)
      });
    }
  };
}

/**
 * @private
 * @param makeRequest - function to make requests via an adapter
 * @param data - Raw role data
 * @return Wrapped role data
 */
export function wrapRole(makeRequest, data) {
  const role = toPlainObject(copy(data));
  const roleWithMethods = enhanceWithMethods(role, createRoleApi(makeRequest));
  return freezeSys(roleWithMethods);
}

/**
 * @private
 */
export const wrapRoleCollection = wrapCollection(wrapRole);
