// contentful-management/dist/es-modules/entities/space-membership.js
import copy from 'fast-copy';
import { freezeSys, toPlainObject } from 'contentful-sdk-core';
import enhanceWithMethods from '../enhance-with-methods';
import { wrapCollection } from '../common-utils';
/**
 * @private
 */
function createSpaceMembershipApi(makeRequest) {
  const getParams = data => ({
    spaceId: data.sys.space.sys.id,
    spaceMembershipId: data.sys.id
  });
  return {
    update: function update() {
      const data = this.toPlainObject();
      return makeRequest({
        entityType: 'SpaceMembership',
        action: 'update',
        params: getParams(data),
        payload: data
      }).then(data => wrapSpaceMembership(makeRequest, data));
    },
    delete: function del() {
      const data = this.toPlainObject();
      return makeRequest({
        entityType: 'SpaceMembership',
        action: 'delete',
        params: getParams(data)
      });
    }
  };
}

/**
 * @private
 * @param makeRequest - function to make requests via an adapter
 * @param data - Raw space membership data
 * @return Wrapped space membership data
 */
export function wrapSpaceMembership(makeRequest, data) {
  const spaceMembership = toPlainObject(copy(data));
  const spaceMembershipWithMethods = enhanceWithMethods(spaceMembership, createSpaceMembershipApi(makeRequest));
  return freezeSys(spaceMembershipWithMethods);
}

/**
 * @private
 */
export const wrapSpaceMembershipCollection = wrapCollection(wrapSpaceMembership);
