// contentful-management/dist/es-modules/entities/workflow.js
function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }
function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { _defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }
function _defineProperty(obj, key, value) { key = _toPropertyKey(key); if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }
function _toPropertyKey(arg) { var key = _toPrimitive(arg, "string"); return typeof key === "symbol" ? key : String(key); }
function _toPrimitive(input, hint) { if (typeof input !== "object" || input === null) return input; var prim = input[Symbol.toPrimitive]; if (prim !== undefined) { var res = prim.call(input, hint || "default"); if (typeof res !== "object") return res; throw new TypeError("@@toPrimitive must return a primitive value."); } return (hint === "string" ? String : Number)(input); }
import { freezeSys, toPlainObject } from 'contentful-sdk-core';
import copy from 'fast-copy';
import { wrapCollection } from '../common-utils';
import enhanceWithMethods from '../enhance-with-methods';
/**
 * @private
 */
export default function createWorkflowApi(makeRequest) {
  const getParams = Workflow => ({
    spaceId: Workflow.sys.space.sys.id,
    environmentId: Workflow.sys.environment.sys.id,
    workflowId: Workflow.sys.id
  });
  return {
    update: function update() {
      const raw = this.toPlainObject();
      return makeRequest({
        entityType: 'Workflow',
        action: 'update',
        params: getParams(raw),
        payload: raw
      }).then(data => wrapWorkflow(makeRequest, data));
    },
    delete: function _delete() {
      const raw = this.toPlainObject();
      return makeRequest({
        entityType: 'Workflow',
        action: 'delete',
        params: _objectSpread(_objectSpread({}, getParams(raw)), {}, {
          version: raw.sys.version
        })
      }).then(() => {
        // noop
      });
    },
    complete: function complete() {
      const raw = this.toPlainObject();
      return makeRequest({
        entityType: 'Workflow',
        action: 'complete',
        params: _objectSpread(_objectSpread({}, getParams(raw)), {}, {
          version: raw.sys.version
        })
      }).then(() => {
        // noop
      });
    }
  };
}

/**
 * @private
 */
export function wrapWorkflow(makeRequest, data) {
  const Workflow = toPlainObject(copy(data));
  const WorkflowWithMethods = enhanceWithMethods(Workflow, createWorkflowApi(makeRequest));
  return freezeSys(WorkflowWithMethods);
}

/**
 * @private
 */
export const wrapWorkflowCollection = wrapCollection(wrapWorkflow);
