// contentful-management/dist/es-modules/adapters/REST/endpoints/entry.js
function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }
function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { _defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }
function _defineProperty(obj, key, value) { key = _toPropertyKey(key); if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }
function _toPropertyKey(arg) { var key = _toPrimitive(arg, "string"); return typeof key === "symbol" ? key : String(key); }
function _toPrimitive(input, hint) { if (typeof input !== "object" || input === null) return input; var prim = input[Symbol.toPrimitive]; if (prim !== undefined) { var res = prim.call(input, hint || "default"); if (typeof res !== "object") return res; throw new TypeError("@@toPrimitive must return a primitive value."); } return (hint === "string" ? String : Number)(input); }
import copy from 'fast-copy';
import * as raw from './raw';
import { normalizeSelect } from './utils';
export const get = (http, params, rawData, headers) => {
  return raw.get(http, `/spaces/${params.spaceId}/environments/${params.environmentId}/entries/${params.entryId}`, {
    params: normalizeSelect(params.query),
    headers: _objectSpread({}, headers)
  });
};
export const getPublished = (http, params, rawData, headers) => {
  return raw.get(http, `/spaces/${params.spaceId}/environments/${params.environmentId}/public/entries`, {
    params: normalizeSelect(params.query),
    headers: _objectSpread({}, headers)
  });
};
export const getMany = (http, params, rawData, headers) => {
  return raw.get(http, `/spaces/${params.spaceId}/environments/${params.environmentId}/entries`, {
    params: normalizeSelect(params.query),
    headers: _objectSpread({}, headers)
  });
};
export const patch = (http, params, data, headers) => {
  return raw.patch(http, `/spaces/${params.spaceId}/environments/${params.environmentId}/entries/${params.entryId}`, data, {
    headers: _objectSpread({
      'X-Contentful-Version': params.version,
      'Content-Type': 'application/json-patch+json'
    }, headers)
  });
};
export const update = (http, params, rawData, headers) => {
  var _rawData$sys$version;
  const data = copy(rawData);
  delete data.sys;
  return raw.put(http, `/spaces/${params.spaceId}/environments/${params.environmentId}/entries/${params.entryId}`, data, {
    headers: _objectSpread({
      'X-Contentful-Version': (_rawData$sys$version = rawData.sys.version) !== null && _rawData$sys$version !== void 0 ? _rawData$sys$version : 0
    }, headers)
  });
};
export const del = (http, params) => {
  return raw.del(http, `/spaces/${params.spaceId}/environments/${params.environmentId}/entries/${params.entryId}`);
};
export const publish = (http, params, rawData) => {
  return raw.put(http, `/spaces/${params.spaceId}/environments/${params.environmentId}/entries/${params.entryId}/published`, null, {
    headers: {
      'X-Contentful-Version': rawData.sys.version
    }
  });
};
export const unpublish = (http, params) => {
  return raw.del(http, `/spaces/${params.spaceId}/environments/${params.environmentId}/entries/${params.entryId}/published`);
};
export const archive = (http, params) => {
  return raw.put(http, `/spaces/${params.spaceId}/environments/${params.environmentId}/entries/${params.entryId}/archived`);
};
export const unarchive = (http, params) => {
  return raw.del(http, `/spaces/${params.spaceId}/environments/${params.environmentId}/entries/${params.entryId}/archived`);
};
export const create = (http, params, rawData) => {
  const data = copy(rawData);
  return raw.post(http, `/spaces/${params.spaceId}/environments/${params.environmentId}/entries`, data, {
    headers: {
      'X-Contentful-Content-Type': params.contentTypeId
    }
  });
};
export const createWithId = (http, params, rawData) => {
  const data = copy(rawData);
  return raw.put(http, `/spaces/${params.spaceId}/environments/${params.environmentId}/entries/${params.entryId}`, data, {
    headers: {
      'X-Contentful-Content-Type': params.contentTypeId
    }
  });
};
export const references = (http, params) => {
  const spaceId = params.spaceId,
    environmentId = params.environmentId,
    entryId = params.entryId,
    include = params.include;
  const level = include || 2;
  return raw.get(http, `/spaces/${spaceId}/environments/${environmentId}/entries/${entryId}/references?include=${level}`);
};
